File: js/src/typed/nodes/StretchNode.ts
// Copyright (c) 2018 Spotify AB.
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

/* Generated */

import { v4 as uuid } from 'uuid';
import { Command, Node, ContentType } from '../../score';
import ArgMapper from '../../schema/ArgMapper';
import TypedNode from './TypedNode';
import AudioParam from '../params/AudioParam';

/**
 * StretchNode.
 */
class StretchNode extends TypedNode {
  /**
   * Unique identifier for the plugin kind this node represents.
   */
  public static PLUGIN_KIND = 'com.nativeformat.plugin.time.stretch';

  /**
   * `pitchRatio` param factory.
   */
  private static PITCH_RATIO_PARAM = AudioParam.newParamMapper('pitchRatio', 1);

  /**
   * `stretch` param factory.
   */
  private static STRETCH_PARAM = AudioParam.newParamMapper('stretch', 1);

  /**
   * `formantRatio` param factory.
   */
  private static FORMANT_RATIO_PARAM = AudioParam.newParamMapper(
    'formantRatio',
    1
  );

  /**
   * Creates a new `StretchNode` instance.
   */
  constructor(
    id: string,
    public readonly pitchRatio: AudioParam,
    public readonly stretch: AudioParam,
    public readonly formantRatio: AudioParam
  ) {
    super(id, StretchNode.PLUGIN_KIND);
  }

  /**
   * `StretchNode` factory.
   */
  public static create(id: string = uuid()): StretchNode {
    return new StretchNode(
      id,
      StretchNode.PITCH_RATIO_PARAM.create(),
      StretchNode.STRETCH_PARAM.create(),
      StretchNode.FORMANT_RATIO_PARAM.create()
    );
  }

  /**
   * Creates a new `StretchNode` from a score `Node`.
   */
  public static from(node: Node): StretchNode {
    if (StretchNode.PLUGIN_KIND !== node.kind) {
      throw new Error(`Expected plugin kind="${StretchNode.PLUGIN_KIND}"`);
    }
    return new StretchNode(
      node.id,
      StretchNode.PITCH_RATIO_PARAM.readParam(node),
      StretchNode.STRETCH_PARAM.readParam(node),
      StretchNode.FORMANT_RATIO_PARAM.readParam(node)
    );
  }

  /**
   * Params for this node.
   */
  public getParams(): { [key: string]: Command[] } {
    return {
      pitchRatio: this.pitchRatio.getCommands(),
      stretch: this.stretch.getCommands(),
      formantRatio: this.formantRatio.getCommands()
    };
  }

  /**
   * Inputs for this node.
   */
  public getInputs(): Map<string, ContentType> {
    return new Map([['audio', ContentType.AUDIO]]);
  }

  /**
   * Outputs for this node.
   */
  public getOutputs(): Map<string, ContentType> {
    return new Map([['audio', ContentType.AUDIO]]);
  }
}

/**
 * Export.
 */
export default StretchNode;

File: java/src/main/java/com/spotify/nativeformat/typed/nodes/StretchNode.java
/*-
 * -\-\-
 * nf-grapher-java
 * --
 * Copyright (C) 2016 - 2018 Spotify AB
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * -/-/-
 */

/* Generated */

package com.spotify.nativeformat.typed.nodes;

import com.spotify.nativeformat.schema.ParamMapper;
import com.spotify.nativeformat.score.Command;
import com.spotify.nativeformat.score.ContentType;
import com.spotify.nativeformat.score.LoadingPolicy;
import com.spotify.nativeformat.score.Node;
import com.spotify.nativeformat.typed.params.AudioParam;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** A plugin that can independently stretch time and shift the pitch of an audio stream. */
public class StretchNode extends TypedNode {

  /** Unique identifier for the plugin kind this node represents. */
  public static final String PLUGIN_KIND = "com.nativeformat.plugin.time.stretch";

  private static final ParamMapper<AudioParam> PITCH_RATIO_PARAM =
      AudioParam.newParamMapper("pitchRatio", 1D);

  private static final ParamMapper<AudioParam> STRETCH_PARAM =
      AudioParam.newParamMapper("stretch", 1D);

  private static final ParamMapper<AudioParam> FORMANT_RATIO_PARAM =
      AudioParam.newParamMapper("formantRatio", 1D);

  private AudioParam pitchRatio;

  private AudioParam stretch;

  private AudioParam formantRatio;

  private StretchNode(
      String id,
      LoadingPolicy loadingPolicy,
      AudioParam pitchRatio,
      AudioParam stretch,
      AudioParam formantRatio) {
    super(id, PLUGIN_KIND, loadingPolicy);

    this.pitchRatio = pitchRatio;

    this.stretch = stretch;

    this.formantRatio = formantRatio;
  }

  /**
   * Factory method. Generates a random id for the created instance.
   *
   * @return a new StretchNode
   */
  public static StretchNode create() {
    return create(null, LoadingPolicy.ALL_CONTENT_PLAYTHROUGH);
  }

  /**
   * Factory method. Generates a random id for the created instance.
   *
   * @param loadingPolicy the desired node loading policy
   * @return a new StretchNode
   */
  public static StretchNode create(LoadingPolicy loadingPolicy) {
    return create(null, loadingPolicy);
  }

  /**
   * Factory method.
   *
   * @param id the desired node id
   * @param loadingPolicy the desired node loading policy
   * @return a new StretchNode
   */
  public static StretchNode create(String id, LoadingPolicy loadingPolicy) {
    return new StretchNode(
        id,
        loadingPolicy,
        PITCH_RATIO_PARAM.create(),
        STRETCH_PARAM.create(),
        FORMANT_RATIO_PARAM.create());
  }

  /**
   * Creates a new StretchNode from the given Score Node.
   *
   * @param node the Score Node to convert from
   * @return a new StretchNode
   */
  public static StretchNode from(Node node) {
    if (!PLUGIN_KIND.equals(node.kind())) {
      throw new RuntimeException("expected plugin kind=" + PLUGIN_KIND);
    }

    return new StretchNode(
        node.id(),
        node.loadingPolicy(),
        PITCH_RATIO_PARAM.readParam(node),
        STRETCH_PARAM.readParam(node),
        FORMANT_RATIO_PARAM.readParam(node));
  }

  /**
   * An audio parameter specifying the pitch multiplier. For example, a pitchRatio value of 2.0 will
   * double the original audio frequencies.
   *
   * @return AudioParam
   */
  public AudioParam pitchRatio() {
    return this.pitchRatio;
  }

  /**
   * An audio parameter specifying the time stretch multiplier. For example, a stretch value of 2.0
   * will make audio play at half the original speed.
   *
   * @return AudioParam
   */
  public AudioParam stretch() {
    return this.stretch;
  }

  /**
   * An audio parameter specifying the formant envelope multiplier. If the formantRatio is equal to
   * the pitchRatio, formant-preserving pitch shifting will be performed.
   *
   * @return AudioParam
   */
  public AudioParam formantRatio() {
    return this.formantRatio;
  }

  @Override
  public Map<String, List<Command>> params() {
    final Map<String, List<Command>> paramsResult = new HashMap<>();
    PITCH_RATIO_PARAM.addToMap(pitchRatio, paramsResult);
    STRETCH_PARAM.addToMap(stretch, paramsResult);
    FORMANT_RATIO_PARAM.addToMap(formantRatio, paramsResult);
    return paramsResult;
  }

  @Override
  public Map<String, Object> config() {
    final Map<String, Object> configResult = new HashMap<>();

    return configResult;
  }

  @Override
  public Map<String, ContentType> inputs() {
    final Map<String, ContentType> inputsResult = new HashMap<>();
    inputsResult.put("audio", ContentType.AUDIO);
    return inputsResult;
  }

  @Override
  public Map<String, ContentType> outputs() {
    final Map<String, ContentType> outputsResult = new HashMap<>();
    outputsResult.put("audio", ContentType.AUDIO);
    return outputsResult;
  }
}
